### PR TITLE
add LatestView to checkpoint block info, refresh when starting fast hot stuff event loop, and use this value to override tip's view

### PR DESCRIPTION
**Pull request by lazynina**

### CODE REVIEW

**Comment by lazynina** [comment]
* **#1213** <a href="https://app.graphite.dev/github/pr/deso-protocol/core/1213?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `feature/proof-of-stake`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @lazynina and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: consensus/event_loop.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ func (fc *fastHotStuffEventLoop) Init(
 	genesisQC QuorumCertificate,
 	tip BlockWithValidatorList,
 	safeBlocks []BlockWithValidatorList,
+	checkpointLatestView uint64,
```
Comment (tholonious) [inline consensus/event_loop.go:48]: Please decouple the concept of a checkpoint from the event loop. May as well make this param the currentView the event loop initializes with so it's not another override
```suggestion
	currentView uint64,
```

File: consensus/event_loop.go [hunk 2]
```diff
@@ -71,8 +72,12 @@ func (fc *fastHotStuffEventLoop) Init(
 		return errors.Wrap(err, "FastHotStuffEventLoop.Init: ")
 	}
 
-	// We track the current view here so we know which view to time out on later on.
 	fc.currentView = tip.Block.GetView() + 1
+	// If we have a checkpoint latest view from the blockchain that is greater than our tip's view,
+	// we update the current view to the checkpoint's latest view.
+	if checkpointLatestView > fc.currentView {
+		fc.currentView = checkpointLatestView
+	}
```
Comment (tholonious) [inline consensus/event_loop.go:80]: You can also move this up to line 66 so the validation happens before the tip and safe blocks are stored
```suggestion
	// The currentView must be higher than the tip block's current view
	if currentView < tip.GetView() + 1 {
		return errors.new("FastHotStuffEventLoop.Init: currentView is lower than the tip block's view")
	}
	fc.currentView = currentView
```

File: lib/pos_consensus.go [hunk 3]
```diff
@@ -109,9 +109,24 @@ func (fc *FastHotStuffConsensus) Start() error {
...
 	timeoutBaseDuration := time.Millisecond * time.Duration(currentSnapshotGlobalParams.TimeoutIntervalMillisecondsPoS)
 
+	// Refresh the checkpoint block info, so we can get tha latest view.
+	fc.blockchain.updateCheckpointBlockInfo()
+	checkpointBlockInfo := fc.blockchain.GetCheckpointBlockInfo()
+	currentView := uint64(0)
+	if checkpointBlockInfo != nil {
+		currentView = checkpointBlockInfo.LatestView
+	}
+
```
Comment (tholonious) [inline lib/pos_consensus.go:119]: You can do the view comparison here so it's encapsulated here where we perform the checkpoint request
```suggestion
	currentView := tipBlock.Header.GetView() + 1
	if checkpointBlockInfo != nil && checkpointBlockInfo.LatestView > currentView {
		currentView = checkpointBlockInfo.LatestView
	}
```

**Review by tholonious — CHANGES_REQUESTED** [review]

**Review by tholonious — APPROVED** [review]

**Pull request merged**